tools/codegen/api/autograd.py

```diff
@@ -335,9 +335,34 @@ def repl(m: Match[str]) -> str:
                     required_primals = required_primals + ("self",) if required_primals else ("self",)
 
                 if not is_exact_match:
-                    # Make sure that the forward grad is modified inplace when the original formula
-                    # is out of place
-                    formula = f"self_t_raw.defined() ? self_t_raw.copy_({formula}) : {formula}"
+                    # NOTE [In-place forward AD formula Optimization]
+                    #
+                    # This optimization transforms the formula to directly do inplace, i.e.
+                    # instead of self_t.copy_(self_t.op()) we do self_t.op_() when the following are met:
+                    #
+                    # 1) the formula satisfies the pattern: "self_t.op(*args)"
+                    # 2) "op" in (1) needs to be the same as the op the derivative is for
+                    #
+                    # (2) may seem too strict, but currently the only ops that satisfy (1) also satisfy (2)
+                    # If there is a need, we can relax (2) to allow any op that has an in-place variant
+                    # We can also avoid hard coding self_t, if there is a need, but that is more work.
+                    is_single_method_on_self_t = False
+                    match = re.fullmatch(r'self_t.([\w]*)\((.*)\)', formula)
```

Comment:
I think you want to add a `^` at the beginning and `$` at the end of the regex to make sure that you don't have something like `self_t.foo(bar) + self_t.bar(baz)` that matches.

Reply:
Yup, but I used fullmatch instead of match here, so that should already apply automatically?
```
>>> import re
>>> re.fullmatch(r"a", "ab")
>>> re.fullmatch(r"a", "ab") is None
True
>>> re.match(r"a", "ab") is None
False
>>> 
```